packages/firefly/src/AppRouter.tsx

```diff
@@ -0,0 +1,188 @@
+import { useAreModulesReady, useAreModulesRegistered } from "@squide/webpack-module-federation";
+import { useEffect, useMemo, useState, type ReactNode } from "react";
+import { Outlet, RouterProvider, createBrowserRouter, useLocation, type RouterProviderProps } from "react-router-dom";
+
+export type OnLoadPublicDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type onLoadProtectedDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type OnCompleteRegistrationFunction = () => Promise<unknown>;
+
+interface RootRouteProps {
+    fallback: ReactNode;
+    onLoadPublicData?: OnLoadPublicDataFunction;
+    onLoadProtectedData?: onLoadProtectedDataFunction;
+    onCompleteRegistration?: OnCompleteRegistrationFunction;
+    waitForMsw: boolean;
+    areModulesRegistered: boolean;
+    areModulesReady: boolean;
+}
+
+// Most of the bootstrapping logic has been moved to this component because AppRouter
+// cannot leverage "useLocation" since it must be used in a child component of "RouterProvider".
+export function RootRoute(props: RootRouteProps) {
+    const {
+        fallback,
+        onLoadPublicData,
+        onLoadProtectedData,
+        onCompleteRegistration,
+        waitForMsw,
+        areModulesRegistered,
+        areModulesReady
+    } = props;
+
+    // Initialize as loaded if no handler is provided to load the public data.
+    const [isPublicDataLoaded, setIsPublicDataLoaded] = useState(!onLoadPublicData);
+
+    // Initialize as loaded if no handler is provided to load the protected data.
+    const [isProtectedDataLoaded, setIsProtectedDataLoaded] = useState(!onLoadProtectedData);
+
+    const logger = useLogger();
+    const location = useLocation();
+
+    // Re-render the app once MSW is started, otherwise, the API calls for module routes will return a 404 status.
+    const isMswStarted = useIsMswStarted(waitForMsw);
+
+    useEffect(() => {
+        // Only log these messages if MSW is enabled.
+        if (waitForMsw) {
+            if ((areModulesRegistered || areModulesReady) && !isMswStarted) {
+                logger.debug(`[shell] Modules are ${areModulesReady ? "ready" : "registered"}, waiting for MSW to start...`);
+            }
+
+            if (!areModulesRegistered && !areModulesReady && isMswStarted) {
+                logger.debug("[shell] MSW is started, waiting for the modules...");
+            }
+        }
+    }, [logger, areModulesRegistered, areModulesReady, isMswStarted, waitForMsw]);
+
+    useEffect(() => {
+        // Don't go further if no handler has been provided to load public data.
+        if (onLoadPublicData) {
+            if ((areModulesRegistered || areModulesReady) && isMswStarted) {
+                if (!isPublicDataLoaded) {
+                    logger.debug("[shell] Loading public data.");
+
+                    const abordController = new AbortController();
+
+                    onLoadPublicData(abordController.signal).finally(() => {
```

Comment:
onLoadPublicData(abordController.signal).finally

Using the finally means that even if there was an error while loading the data, we still set setIsPublicDataLoaded to true, and write "Public data has been loaded". Shouldn't put this in the ".then" and have something for when there is an error?

This could be done in a separate PR if people are waiting on the AppRouter

Reply:
Indeed, I am rewriting this part as part of the error handling code that I am currently adding to AppRouter.